### PR TITLE
refactor: register type commands only when needed

### PR DESCRIPTION
**Pull request by xiyaowong**
At present, the problem in #1921 seems to be unavoidable. The current approach is that when composite keys are not used, the logic is the same as before. See if users can accept it  @theol0403 @wenfangdu @vlwkaos

### CODE REVIEW

**Comment by wenfangdu** [comment]
@xiyaowong Bro, leave a `.vsix` file, so that we can try.

**Comment by xiyaowong** [comment]
@wenfangdu https://github.com/vscode-neovim/vscode-neovim/actions/runs/8927691229

**Comment by wenfangdu** [comment]
@xiyaowong You mean get `.vsix` from that page?

**Comment by xiyaowong** [comment]
@wenfangdu 
image

**Comment by wenfangdu** [comment]
@xiyaowong Seems normal to me:

https://github.com/vscode-neovim/vscode-neovim/assets/28700378/0c1d78e2-e6f9-4cd7-9055-cdfc332f9a8e

**Comment by theol0403** [comment]
Code looks fine. Merge when tested.

**Comment by wenfangdu** [comment]
@xiyaowong Seems like https://github.com/vscode-neovim/vscode-neovim/pull/1923 already fixed https://github.com/vscode-neovim/vscode-neovim/issues/1921, not reproducible in v1.11.1 🤔.

File: src/typing_manager.ts [hunk 1]
```diff
@@ -197,7 +237,7 @@ export class TypingManager implements Disposable {
             const matchedFirstKey = this.compositeMatchedFirstKey;
             this.compositeMatchedFirstKey = undefined;
             const { command, args } = this.compositeKeys[matchedFirstKey + key];
-            await commands.executeCommand(command, ...(args ? args : []));
+            commands.executeCommand(command, ...(args ? args : []));
```
Comment (justinmk) [inline src/typing_manager.ts:240]: why remove `await`? same question for line 325

Comment (xiyaowong) [reply src/typing_manager.ts:240]: As seen in the discussion of vscode related issues, make the execution as lightweight as possible. Whether or not await is added here has no effect on vscode-neovim.

File: src/typing_manager.ts [hunk 2]
```diff
@@ -35,19 +35,62 @@ export class TypingManager implements Disposable {
...
+    }
+
+    private set takeOverVSCodeInput(takeOver: boolean) {
+        this._takeOverVSCodeInput = takeOver;
+
+        if (takeOver) {
+            if (!this.typeHandler) this.typeHandler = commands.registerTextEditorCommand("type", this.onVSCodeType);
+            if (!this.replacePreviousCharHandler)
+                this.replacePreviousCharHandler = commands.registerCommand(
+                    "replacePreviousChar",
```
Comment (justinmk) [inline src/typing_manager.ts:80]: why *register* the command every time `takeOverVSCodeInput=true` is assigned? I would think that the command(s) can be registered once, and there implementation can check state to decide what to do.

Comment (xiyaowong) [reply src/typing_manager.ts:80]: >     // Notes:
>     // 1. "type" and "replacePreviousChar" must be registered at the same time
>     // 2. Forwarding the arguments of replacePreviousChar to default:replacePreviousChar
>     //    causes text jitter during ime composition
>     //
>     // The compromise solution used here:
>     // 1. When composite keys are needed, always register type and
>     //    replacePreviousChar at the same time, which unavoidably causes text
>     //    flickering during ime composition.
>     // 2. When composite keys are not needed, only register type and
>     //    replacePreviousChar when it's necessary to take over vscode input.

Comment (justinmk) [reply src/typing_manager.ts:80]: oh, I see that this is overwriting a vscode builtin: https://github.com/microsoft/vscode/blob/3982027f02dee5aa52622f7f2812c70f04ce337a/src/vs/editor/browser/coreCommands.ts#L2167

it sure would be worth mentioning that in a code comment somewhere.

Comment (justinmk) [reply src/typing_manager.ts:80]: I did read that comment, but it doesn't mention that this is a vscode builtin, which is critical for understanding this.

Comment (xiyaowong) [reply src/typing_manager.ts:80]: We can mention [vim-sample](https://github.com/microsoft/vscode-extension-samples/tree/main/vim-sample) in CONTRIBUTING.md

Comment (justinmk) [reply src/typing_manager.ts:80]: sure, but that still leaves other developers with a puzzle to piece together. It costs nearly zero to mention in this context that "replacePreviousChar" is a vscode builtin.

Comment (xiyaowong) [reply src/typing_manager.ts:80]: Oh, the relevant content has been explained here: https://github.com/vscode-neovim/vscode-neovim/blob/master/CONTRIBUTING.md#typingmanager

Comment (justinmk) [reply src/typing_manager.ts:80]: Again, we are talking about a 1-line addition that is *near* this code.

Comment (xiyaowong) [reply src/typing_manager.ts:80]: What's the relation between this and this PR? More than just these two vscode commands are overridden.

Comment (justinmk) [reply src/typing_manager.ts:80]: This is something I noticed while reviewing this PR. It's a zero-cost improvement that can be made, regardless of other possible, future improvements. Is any other justification needed?

**Pull request merged**